src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs

```diff
@@ -507,11 +505,28 @@ private void WriteResponseHeadersUnsynchronized(int streamId, int statusCode, Ht
     {
         try
         {
+            // In the case of the headers, there is always a status header to be returned, so BeginEncodeHeaders will not return BufferTooSmall.
             _headersEnumerator.Initialize(headers);
             _outgoingFrame.PrepareHeaders(headerFrameFlags, streamId);
-            var buffer = _headerEncodingBuffer.AsSpan();
+            _headerEncodingBuffer.ResetWrittenCount();
+            var buffer = _headerEncodingBuffer.GetSpan(_maxFrameSize)[0.._maxFrameSize]; // GetSpan might return more data that can result in a less deterministic behavior on the way headers are split into frames.
             var done = HPackHeaderWriter.BeginEncodeHeaders(statusCode, _hpackEncoder, _headersEnumerator, buffer, out var payloadLength);
-            FinishWritingHeadersUnsynchronized(streamId, payloadLength, done);
+            Debug.Assert(done != HPackHeaderWriter.HeaderWriteResult.BufferTooSmall, "Oversized frames should not be returned, because this always writes the status.");
+            if (done == HPackHeaderWriter.HeaderWriteResult.Done)
+            {
+                // Fast path, only a single HEADER frame.
+                _outgoingFrame.PayloadLength = payloadLength;
+                _outgoingFrame.HeadersFlags |= Http2HeadersFrameFlags.END_HEADERS;
+                WriteHeaderUnsynchronized();
+                _outputWriter.Write(buffer[0..payloadLength]);
+            }
+            else
+            {
+                // More headers sent in CONTINUATION frames.
+                _headerEncodingBuffer.Advance(payloadLength);
+                SplitHeaderFramesToOutput(streamId, endOfHeaders: false, isFramePrepared: true);
```

Comment:
Yes, in this branch `BeginEncodeHeaders` returns `MoreHeaders`, so it means we can write the current `buffer` as is to the output and continue writing the rest of the headers with `FinishWritingHeadersUnsynchronized`.

This branch could also write the frame directly as I don't think in this case we ever get a `buffer` larger to a single frame (so we could drop `SplitHeaderFramesToOutput` and write the frame directly as in the `Done` case but without the `END_HEADERS` flag). Note that this is not true when writing the TRAILERS, because that case might return `BufferTooSmall` already for the initial write to the buffer frame. - I am happy to change this branch as I propose here, if you think it worth it.

In this case the frame is prepared as a `HEADERS` frame in line 507, it is sent and further `CONTINUATION` frames will be used by `FinishWritingHeadersUnsynchronized`. I think the comment here might be misleading, I will clarify that.